tests/protocol/test_wire.py

```diff
@@ -1056,6 +1057,109 @@ def test_forced_update_should_update_the_goal_state_and_the_host_plugin_when_the
+                filter_calls = lambda calls, regex=None: (c for c in calls_to_strings(calls) if regex is None or re.match(regex, c))
+                logger_calls = lambda regex=None: [m for m in filter_calls(logger.method_calls, regex)]
+                warnings = lambda: logger_calls(r'call.warn\(.*An error occurred while retrieving the goal state.*')
+                periodic_warnings = lambda: logger_calls(r'call.periodic_warn\(.*Attempts to retrieve the goal state are failing.*')
+                success_messages = lambda: logger_calls(r'call.info\(.*Retrieving the goal state recovered from previous errors.*')
+                telemetry_calls = lambda regex=None: [m for m in filter_calls(add_event.mock_calls, regex)]
+                goal_state_events = lambda: telemetry_calls(r".*op='FetchGoalState'.*")
+
+                #
+                # Initially calls to retrieve the goal state are successful...
+                #
+                fail_goal_state_request = False
+                with create_log_and_telemetry_mocks() as (logger, add_event):
+                    protocol.client.try_update_goal_state()
+
+                    lc = logger_calls()
+                    self.assertTrue(len(lc) == 0, "A successful call should not produce any log messages: [{0}]".format(lc))
+
+                    tc = telemetry_calls()
+                    self.assertTrue(len(tc) == 0, "A successful call should not produce any telemetry events: [{0}]".format(tc))
+
+                #
+                # ... then an error happens...
+                #
+                fail_goal_state_request = True
+                with create_log_and_telemetry_mocks() as (logger, add_event):
+                    protocol.client.try_update_goal_state()
+
+                    w = warnings()
+                    pw = periodic_warnings()
+                    self.assertEquals(len(w), 1, "A failure should have produced a warning: [{0}]".format(w))
+                    self.assertEquals(len(pw), 1, "A failure should have produced a periodic warning: [{0}]".format(pw))
+
+                    gs = goal_state_events()
+                    self.assertTrue(len(gs) == 1 and 'is_success=False' in gs[0], "A failure should produce a telemetry event (success=false): [{0}]".format(gs))
+
+                #
+                # ... and errors continue happening...
+                #
+                with create_log_and_telemetry_mocks() as (logger, add_event):
+                    protocol.client.try_update_goal_state()
+                    protocol.client.try_update_goal_state()
+                    protocol.client.try_update_goal_state()
+
+                    w = warnings()
+                    pw = periodic_warnings()
+                    self.assertTrue(len(w) == 0, "Subsequent failures should not produce warnings: [{0}]".format(w))
+                    self.assertEquals(len(pw), 3, "Subsequent failures should produce periodic warnings: [{0}]".format(pw))
```

Comment:
Why would periodic warnings be 3 instead of 1 if we are logging them every 6 hours?

Reply:
we always call periodic_warning... the underlying implementation should handle the period

Reply:
Got it, this just asserts the periodic is being called.